fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/MetaDataProtoEditor.java

```diff
@@ -0,0 +1,244 @@
+import java.util.function.Function;
+
+/**
+ * A helper class for mutating the meta-data proto.
+ *
+ * <p>
+ * This class contains several helper methods for modifying a serialized meta-data, e.g. add a new record type to the meta-data.
+ * {@link FDBMetaDataStore#mutateMetaData(Function)} is one example of where these methods can be useful. {@code FDBMetaDataStore#mutateMetaData}
+ * modifies the stored meta-data using a mutation callback and saves it back to the meta-data store.
+ * </p>
+ *
+ */
+@API(API.Status.EXPERIMENTAL)
+public class MetaDataProtoEditor {
+    /**
+     * Add a new record type to the meta-data.
+     *
+     * <p>
+     * Adding the record type involves three steps: the message type is added to the list of file descriptor's
+     * message types, a field of the given type is added to the union, and its primary key is set. This method takes
+     * care of the first two. Callers must also set the primary key
+     * ({@link com.apple.foundationdb.record.metadata.RecordTypeBuilder#setPrimaryKey} for the type.
+     * Note that adding {@code UNION} record types is not allowed. To add {@code NESTED} record types use {@link #addNestedRecordType}.
+     * </p>
+     *
+     * @param metaDataBuilder the meta-data builder
+     * @param newRecordType the new record type
+     * @return the modified meta-data
+     */
+    @Nonnull
+    public static RecordMetaDataProto.MetaData addRecordType(@Nonnull RecordMetaDataProto.MetaData.Builder metaDataBuilder, @Nonnull DescriptorProtos.DescriptorProto newRecordType) {
+        RecordMetaDataOptionsProto.RecordTypeOptions.Usage newRecordTypeUsage = getMessageTypeUsage(newRecordType);
+        if (newRecordType.getName().equals(RecordMetaDataBuilder.DEFAULT_UNION_NAME) ||
+                newRecordTypeUsage == RecordMetaDataOptionsProto.RecordTypeOptions.Usage.UNION) {
+            throw new MetaDataException("Adding UNION record type not allowed");
+        }
+        if (newRecordTypeUsage == RecordMetaDataOptionsProto.RecordTypeOptions.Usage.NESTED) {
+            throw new MetaDataException("Use addNestedRecordType for adding NESTED record types");
+        }
+        if (findMessageTypeByName(metaDataBuilder.getRecordsBuilder(), newRecordType.getName()) != null) {
+            throw new MetaDataException("Record type " + newRecordType.getName() + " already exists");
+        }
+        metaDataBuilder.getRecordsBuilder().addMessageType(newRecordType);
+        addFieldToUnion(metaDataBuilder.getRecordsBuilder(), newRecordType);
+        return metaDataBuilder.build();
+    }
+
+    private static void addFieldToUnion(@Nonnull DescriptorProtos.FileDescriptorProto.Builder fileBuilder, @Nonnull DescriptorProtos.DescriptorProto newRecordType) {
+        DescriptorProtos.DescriptorProto.Builder unionBuilder = fetchUnionBuilder(fileBuilder);
+        if (unionBuilder.getOneofDeclCount() > 0) {
+            throw new MetaDataException("Adding record type to oneOf is not allowed");
```

Comment:
Okay, following up from this comment: https://github.com/FoundationDB/fdb-record-layer/pull/435#discussion_r267048317

I think (think) that it should be safe to allow adding new fields to the (one and only) `oneOf`. The exact warning from the protobuf docs is that it can't be distinguished (at parse time) between none of the fields being set and the newly added field being set if one is still using the *old* descriptor. I don't think that issue would affect us as anyone reading from a store where data has been written with the *new* descriptor will get a meta-data version mismatch, and even if they ignored the `checkVersion` check, then the error would be something like, "no record set in union", which I think is...fine.

Reply:
OK, probably lets for now disallow it, and I fix this in a followup pr?

Reply:
I suppose. What's the work involved in adding the field to the `oneOf` if present?

Reply:
Honestly I don't know. I experimented different ways, but all unsuccessful. It looks like at this point, the union descriptor contains all of the `oneof` fields, and `oneof` is just a decl. Will investigate this later.

Reply:
Ah, it looks like one needs to set the `oneof_index` in the field's descriptor proto (to zero, I think, as we allow only zero or one oneofs in the union): https://github.com/protocolbuffers/protobuf/blob/5e0812d4b1f110fa6c1e33c91c9965a10989c41f/src/google/protobuf/descriptor.proto#L205. I suppose this could wait until later and we can disallow it for now:

Reply:
good finding. thanks.

Reply:
Is there an issue to fix this if it isn't going into this PR?

Reply:
File one now: #484.